mlir/include/mlir/Dialect/MPI/IR/MPIOps.td

```diff
@@ -0,0 +1,114 @@
+//===- MPI.td - Message Passing Interface Ops ---------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MPI_OPS
+#define MPI_OPS
+
+include "mlir/Dialect/MPI/IR/MPIBase.td"
+
+class MPI_Op<string mnemonic, list<Trait> traits = []> :
+    Op<MPI_Dialect, mnemonic, traits>;
+
+//===----------------------------------------------------------------------===//
+// InitOp
+//===----------------------------------------------------------------------===//
+
+def MPI_InitOp : MPI_Op<"init", [
+    
+  ]> {
+  let summary = "Initialize the MPI library, equivalent to `MPI_Init(NULL, NULL)`";
+  let description = [{
+    Passing &argc, &argv is not supported currently.
+    Inspecting the return value (error code) is also not supported.
```

Comment:
Like most C libraries, MPI returns an error code indicating whether the call was successful:
```
int ierr;
ierr = MPI_Init(NULL, NULL);
if (ierr) { // non-zero error code is bad.
   // something bad happened
   ...
}
```
So this is so that application codes can act on the return value and do something appropriate.

Reply:
There's an open question here in how to best model these error codes in MLIR, as the exact value of each existing error code is not well-defined at the standard level, and so the error `MPI_ERR_COMM` (invalid communicator) might be encoded differently by different libraries.

I left that out of the proposal in order to avoid this discussion. I felt this was justified as a lot of the MPI code I've seen so far did not inspect return values.

Reply:
I would write that for now -- in case of errors -- the lowerings are expected to assert/abort.

Reply:
> I felt this was justified as a lot of the MPI code I've seen so far did not inspect return values.

The default MPI-installed error handler calls `MPI_Abort` (which is probably why they don't inspect stuff). A decent library will install its own error handler and act on the return value